cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/MetadataDataset.java

```diff
@@ -413,6 +434,78 @@ private Metadata getSnapshotBeforeTime(Id.NamespacedId targetId, long timeMillis
   }
 
   /**
+   * Returns metadata for a given set of entities
+   *
+   * @param targetIds entities for which metadata is required
+   * @return map of entitiyId to set of metadata for that entity
+   */
+  public Set<Metadata> getMetadata(Set<? extends Id.NamespacedId> targetIds) {
+    List<ImmutablePair<byte [], byte []>> fuzzyKeys = new ArrayList<>();
+    for (Id.NamespacedId targetId : targetIds) {
+      fuzzyKeys.add(getFuzzyKeyFor(targetId));
+    }
+
+    // Sort fuzzy keys
+    Collections.sort(fuzzyKeys, FUZZY_KEY_COMPARATOR);
+
+    // Scan using fuzzy filter. Scan returns one row per property.
+    // Group the rows on entityId
+    Multimap<Id.NamespacedId, MetadataEntry> metadataMap = HashMultimap.create();
+    Scanner scan = indexedTable.scan(new Scan(null, null, new FuzzyRowFilter(fuzzyKeys)));
+    try {
+      Row next;
+      while ((next = scan.next()) != null) {
+        MetadataEntry metadataEntry = convertRow(next);
+        if (metadataEntry != null) {
+          metadataMap.put(metadataEntry.getTargetId(), metadataEntry);
+        }
+      }
+    } finally {
+      scan.close();
+    }
+
+    // Create metadata objects for each entity from grouped rows
+    Set<Metadata> metadataSet = new HashSet<>();
+    for (Map.Entry<Id.NamespacedId, Collection<MetadataEntry>> entry : metadataMap.asMap().entrySet()) {
+      Map<String, String> properties = new HashMap<>();
+      Set<String> tags = Collections.emptySet();
+      for (MetadataEntry metadataEntry : entry.getValue()) {
+        if (TAGS_KEY.equals(metadataEntry.getKey())) {
+          tags = splitTags(metadataEntry.getValue());
+        } else {
+          properties.put(metadataEntry.getKey(), metadataEntry.getValue());
+        }
+      }
+      metadataSet.add(new Metadata(entry.getKey(), properties, tags));
+    }
+    return metadataSet;
+  }
+
+  @Nullable
+  private MetadataEntry convertRow(Row row) {
+    byte[] rowKey = row.getRow();
+    String targetType = MdsKey.getTargetType(rowKey);
+    Id.NamespacedId namespaceId = MdsKey.getNamespaceIdFromKey(targetType, rowKey);
```

Comment:
nit: this is a ``namespacedId``. not ``namespaceId``. That method name in ``MdsKey`` should also be changed. Ok if you want to do it in a later PR, but lets open a jira if so?